File: CampusBikes_TreeMap.py
# TC: O(M x N) where M is the length of workers list and N is the length of bikes list. 
# SC: O(D) where D is the height of the treemap that stores distance between all the workers with all the bikes.

from sortedcontainers import SortedDict

class Solution:
    def assignBikes(self, workers: List[List[int]], bikes: List[List[int]]) -> List[int]:
        if not workers or not bikes or len(workers) == 0 or len(bikes) == 0: 
            return 
        
        if len(workers) == 1 and len(bikes) == 1: 
            return [0]
        
        dist_hmap = SortedDict()
        bool_workers = [False] *len(workers)
        bool_bikes = [False] * len(bikes)
        result = [-1]* len(workers)
        
        for i in range(len(workers)):
            for j in range(len(bikes)): 
                distance = abs(workers[i][0] - bikes[j][0]) + abs(workers[i][1] - bikes[j][1])
                if distance in dist_hmap: 
                    temp_list = dist_hmap.get(distance)
                    temp_list.append((i, j))
                    dist_hmap[distance] = temp_list
                else: 
                    dist = (i,j)
                    dist_hmap[distance] = [dist]
        
        assigned = 0
        for dist in dist_hmap.keys(): 
            temp = dist_hmap.get(dist)
            for x,y in temp: 
                if not bool_workers[x] and not bool_bikes[y]:
                    result[x] = y
                    bool_workers[x] = True
                    bool_bikes[y] = True
                    assigned += 1
                
                if assigned == len(workers): 
                    break
        
        return result

File: CampusBikesTreemap.java
import java.util.Scanner;
import java.util.List;
import java.util.TreeMap;
import java.util.ArrayList;

public class CampusBikesTreemap {

        // Greedy - Tree Map - Time O(m*n + log(m*n)) and Space O(m*n)
        // Use only bucket sort in interviews, not tree map

        public int[] assignBikes(int[][] workers, int[][] bikes) {

            //
            int m = workers.length;
            int n = bikes.length;

            // null case
            if(m == 0 || n == 0) {
                return null;
            }

            // build tree map
            TreeMap<Integer, List<int[]>> pairsAtDist = new TreeMap<>();    // O(m*n) space

            // nested loop to calculate manhattan distances
            for(int i = 0; i < m; i++) {        // O(m*n)
                for(int j = 0; j < n; j++) {

                    // call manhattan distance method
                    int dist = manhattanDist(workers[i], bikes[j]);

                    if(!pairsAtDist.containsKey(dist)) {

                        // add dist - empty list pair, if new distance
                        pairsAtDist.put(dist, new ArrayList<>());    // O(log(m*n))
                    }

                    // add pair to appropriate distance list
                    pairsAtDist.get(dist).add(new int[] {i, j});
                }
            }

            //availability boolean arrays
            boolean[] unavailableWorker = new boolean[m];    // O(m) space
            boolean[] unavailableBike = new boolean[n];      // O(n) space

            // result array
            int[] result = new int[m];                      // O(m) space
            int assignCount = 0;

            // Greedily iterate over pairs at distances starting from minimum distance
            // sorted distances is important
            for(int dist: pairsAtDist.keySet()) {       // O(m*n)

                // get list at a distance
                List<int[]> pairs = pairsAtDist.get(dist);

                // local optimal (minimal) distance pair matches leading to global optimal(minimal) assignment
                for(int[] pair: pairs) {

                    int workerIdx = pair[0];
                    int bikeIdx = pair[1];

                    // if both worker and bike in a pair are available
                    if(!unavailableWorker[workerIdx] && !unavailableBike[bikeIdx]) {

                        // assign and increase count of assigned workers
                        result[workerIdx] = bikeIdx;
                        assignCount++;

                        // make them available for future
                        unavailableWorker[workerIdx] = true;
                        unavailableBike[bikeIdx] = true;

                        // if all bikes assigned
                        if(assignCount == n) {

                            // output
                            return result;
                        }
                    }
                }
            }

            // output
            return result;
        }

        public int manhattanDist(int[] worker, int[] bike) {

            // calculate manhattan distance between worker and bike
            return Math.abs(worker[0] - bike[0]) + Math.abs(worker[1] - bike[1]);
        }

        public static void main(String[] args) {

            CampusBikesTreemap obj = new CampusBikesTreemap();

            Scanner scanner = new Scanner(System.in);

            System.out.println("number of workers");
            int m = scanner.nextInt();
            int[][] workers = new int[m][2];

            System.out.println("number of bikes");
            int n = scanner.nextInt();
            int[][] bikes = new int[n][2];

            System.out.println("positions of workers");
            for(int[] worker: workers) {

                worker[0] = scanner.nextInt();
                worker[1] = scanner.nextInt();

                System.out.println(" ");
            }

            System.out.println("positions of bikes");
            for(int[] bike: bikes) {

                bike[0] = scanner.nextInt();
                bike[1] = scanner.nextInt();

                System.out.println(" ");
            }

            int[] answer = obj.assignBikes(workers, bikes);

            System.out.println("Assigned Campus bikes are ");
            for(int bike: answer) {

                System.out.print(bike + " ");
            }


        }

}

/*
Time Complexity = O(m*n + log(m*n))
Space Complexity = O(m*n)

Input: workers = [[0,0],[2,1]], bikes = [[1,2],[3,3]]
Output: [1,0]
*/